File: apps/mobile/lib/features/invoice/domain/invoice.dart
enum InvoiceStatus {
  unpaid('UNPAID'),
  pendingVerification('PENDING_VERIFICATION'),
  paid('PAID'),
  overdue('OVERDUE'),
  waived('WAIVED');

  const InvoiceStatus(this.apiValue);

  final String apiValue;

  static InvoiceStatus fromApi(Object? value) {
    if (value is! String) {
      throw const FormatException('Invoice status is invalid.');
    }
    for (final status in values) {
      if (status.apiValue == value) return status;
    }
    throw FormatException('Unsupported invoice status: $value');
  }
}

String invoiceStatusLabel(InvoiceStatus status) {
  switch (status) {
    case InvoiceStatus.unpaid:
      return 'Belum Dibayar';
    case InvoiceStatus.pendingVerification:
      return 'Menunggu Verifikasi';
    case InvoiceStatus.paid:
      return 'Lunas';
    case InvoiceStatus.overdue:
      return 'Terlambat';
    case InvoiceStatus.waived:
      return 'Dibebaskan';
  }
}

/// Mirrors `Invoice` in `packages/contracts/src/invoice.ts`.
class Invoice {
  const Invoice({
    required this.id,
    required this.duesTypeId,
    required this.duesTypeName,
    required this.period,
    required this.amount,
    required this.dueDate,
    required this.status,
    required this.houseCode,
    required this.householdDisplayName,
    required this.waivedReason,
    required this.paidAt,
    required this.receiptNumber,
    required this.createdAt,
  });

  final String id;
  final String duesTypeId;
  final String duesTypeName;
  final String period;
  final int amount;

  /// YYYY-MM-DD, matching the API's date-only contract.
  final String dueDate;
  final InvoiceStatus status;
  final String houseCode;
  final String householdDisplayName;
  final String? waivedReason;
  final DateTime? paidAt;
  final String? receiptNumber;
  final DateTime createdAt;

  bool get canSubmitPayment =>
      status == InvoiceStatus.unpaid || status == InvoiceStatus.overdue;

  factory Invoice.fromJson(Map<String, dynamic> json) {
    final paidAtValue = json['paidAt'];
    return Invoice(
      id: json['id'] as String,
      duesTypeId: json['duesTypeId'] as String,
      duesTypeName: json['duesTypeName'] as String,
      period: json['period'] as String,
      amount: (json['amount'] as num).toInt(),
      dueDate: json['dueDate'] as String,
      status: InvoiceStatus.fromApi(json['status']),
      houseCode: json['houseCode'] as String,
      householdDisplayName: json['householdDisplayName'] as String,
      waivedReason: json['waivedReason'] as String?,
      paidAt: paidAtValue == null
          ? null
          : DateTime.parse(paidAtValue as String),
      receiptNumber: json['receiptNumber'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

File: packages/contracts/src/invoice.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const invoiceStatusSchema = z.enum([
  "UNPAID",
  "PENDING_VERIFICATION",
  "PAID",
  "OVERDUE",
  "WAIVED",
]);

export type InvoiceStatus = z.infer<typeof invoiceStatusSchema>;

export const generateInvoicesInputSchema = z
  .object({
    duesTypeId: z.string().uuid(),
    period: z
      .string()
      .regex(/^\d{4}-(0[1-9]|1[0-2])$/, "Periode harus menggunakan format YYYY-MM."),
    amount: z.number().int().positive(),
    dueDate: z
      .string()
      .regex(/^\d{4}-\d{2}-\d{2}$/, "Tanggal harus menggunakan format YYYY-MM-DD."),
  })
  .strict();

export type GenerateInvoicesInput = z.infer<typeof generateInvoicesInputSchema>;

export const waiveInvoiceInputSchema = z
  .object({
    reason: z.string().trim().min(3).max(500),
  })
  .strict();

export type WaiveInvoiceInput = z.infer<typeof waiveInvoiceInputSchema>;

export const invoiceListQuerySchema = z.object({
  status: invoiceStatusSchema.optional(),
  limit: z.coerce.number().int().min(1).max(100).default(20),
});

export const invoiceSchema = z.object({
  id: z.string().uuid(),
  duesTypeId: z.string().uuid(),
  duesTypeName: z.string().min(1),
  period: z.string().min(1),
  amount: z.number().int().nonnegative(),
  dueDate: z.string(),
  status: invoiceStatusSchema,
  houseCode: z.string().min(1),
  householdDisplayName: z.string().min(1),
  waivedReason: z.string().nullable(),
  paidAt: z.string().datetime({ offset: true }).nullable(),
  receiptNumber: z.string().nullable(),
  createdAt: z.string().datetime({ offset: true }),
});

export type Invoice = z.infer<typeof invoiceSchema>;

export const invoiceListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(invoiceSchema) }),
);

export const invoiceDetailResponseSchema = dataEnvelopeSchema(z.object({ invoice: invoiceSchema }));

export const invoiceMutationResponseSchema = dataEnvelopeSchema(
  z.object({ invoice: invoiceSchema }),
);

export const generateInvoicesResponseSchema = dataEnvelopeSchema(
  z.object({ createdCount: z.number().int().nonnegative() }),
);

export type InvoiceListResponse = z.infer<typeof invoiceListResponseSchema>;
export type InvoiceDetailResponse = z.infer<typeof invoiceDetailResponseSchema>;
export type InvoiceMutationResponse = z.infer<typeof invoiceMutationResponseSchema>;
export type GenerateInvoicesResponse = z.infer<typeof generateInvoicesResponseSchema>;
